cpp/src/io/comp/nvcomp_adapter.cpp

```diff
@@ -390,18 +359,109 @@ void batched_compress(compression_type compression,
   update_compression_results(actual_compressed_data_sizes, results, stream);
 }
 
-bool is_compression_enabled(compression_type compression)
+feature_status_parameters::feature_status_parameters()
+  : lib_major_version{NVCOMP_MAJOR_VERSION},
+    lib_minor_version{NVCOMP_MINOR_VERSION},
+    lib_patch_version{NVCOMP_PATCH_VERSION},
+    are_all_integrations_enabled{detail::nvcomp_integration::is_all_enabled()},
+    are_stable_integrations_enabled{detail::nvcomp_integration::is_stable_enabled()}
+{
+  int device;
+  CUDF_CUDA_TRY(cudaGetDevice(&device));
+  CUDF_CUDA_TRY(
+    cudaDeviceGetAttribute(&compute_capability, cudaDevAttrComputeCapabilityMajor, device));
+}
+
+std::optional<std::string> is_compression_disabled(compression_type compression,
+                                                   feature_status_parameters params)
 {
   switch (compression) {
-    case compression_type::DEFLATE:
-      // See https://github.com/rapidsai/cudf/issues/11812
-      return false;
-    case compression_type::SNAPPY: return detail::nvcomp_integration::is_stable_enabled();
-    case compression_type::ZSTD:
-      return NVCOMP_HAS_ZSTD_COMP and detail::nvcomp_integration::is_stable_enabled();
-    default: return false;
+    case compression_type::DEFLATE: {
+      if (not NVCOMP_HAS_DEFLATE(
```

Comment:
I suppose this expands to a valid runtime expression as well as being useful at compile time in `#if` macros? Not super familiar with using macros like this.

Reply:
It is simply a compile time text substitution, so this one will translate to `if (not (MAJOR > 2 or (MAJOR == 2 and MINOR >= 5)))`

Reply:
> Not super familiar with using macros like this.

Yeah, this is tech from the dark ages, thankfully not commonly used any more. I unfortunately could not avoid it here :(

I think you have the right idea, this is pretty much a type-unsafe constexpr function that can be used in `#if` conditions. By the means of text substitution, as Mike said :D